crates/cairo-addons/src/vm/vm_consts.rs

```diff
@@ -0,0 +1,920 @@
+///   access to struct members, pointers, and arrays
+/// - **Introspection**: The dictionary supports Python's `dir()` function to list available
+///   variables
+///
+/// # Implementation Details
+///
+/// The dictionary stores Python objects rather than raw Cairo variables to support
+/// different representations based on the variable type:
+///
+/// - Simple felt values are stored as Python integers
+/// - Relocatable values are stored as `PyRelocatable` objects
+/// - Structs and pointers are stored as `PyVmConst` objects
+///
+/// This approach allows for a more natural Python interface while maintaining
+/// type awareness and memory access capabilities.
+///
+/// # Usage in Hints
+///
+/// In Python hints, this dictionary is available as the `ids` object:
+///
+/// ```python
+/// # Access variables
+/// x = ids.x
+/// y = ids.y
+///
+/// # List available variables
+/// print(dir(ids))
+///
+/// # Check if a variable exists
+/// if hasattr(ids, 'my_var'):
+///     print(f"my_var exists: {ids.my_var}")
+/// ```
+#[pyclass(name = "VmConstsDict", unsendable)]
+pub struct PyVmConstsDict {
+    /// Map of variable names to Python objects
+    pub(crate) items: HashMap<String, Py<PyAny>>,
+}
+
+#[pymethods]
+impl PyVmConstsDict {
+    /// Get a variable by name
+    pub fn __getattr__(&self, name: &str, py: Python<'_>) -> PyResult<PyObject> {
+        match self.items.get(name) {
+            Some(var) => {
+                // Extract the Python object - this could be a PyVmConst or a native Python type
+                // like int
+                Ok(var.clone_ref(py).into_bound_py_any(py)?.into())
+            }
+            None => Err(PyErr::new::<pyo3::exceptions::PyAttributeError, _>(format!(
+                "'VmConstsDict' object has no attribute '{}'",
+                name
+            ))),
+        }
+    }
+
+    /// Get a variable by dictionary access
+    pub fn __getitem__(&self, name: &str, py: Python<'_>) -> PyResult<PyObject> {
+        self.__getattr__(name, py)
+    }
+
+    /// Add or update a variable - accepts any Python object
+    pub fn set_item(&mut self, key: &str, value: Py<PyAny>) {
+        self.items.insert(key.to_string(), value);
+    }
+
+    /// Get all keys
+    pub fn keys(&self, py: Python<'_>) -> PyResult<PyObject> {
+        let keys = PyList::new(py, self.items.keys().map(|k| k.as_str()))?;
+        Ok(keys.into())
+    }
+
+    /// Used for dir() in Python
+    pub fn __dir__(&self, py: Python<'_>) -> PyResult<PyObject> {
+        self.keys(py)
+    }
+
+    /// String representation
+    pub fn __str__(&self) -> String {
+        format!("VmConstsDict with {} items", self.items.len())
+    }
+
+    /// Repr string
+    pub fn __repr__(&self) -> String {
+        format!("VmConstsDict({:?})", self.items.keys().collect::<Vec<_>>())
+    }
+}
+
+/// Creates a VmConstsDict from IDs data and VM
+pub fn create_vm_consts_dict(
```

Comment:
This is pretty complex. 
I am sure some refactoring could be done to improve readability and maintenance.